src/pdn/src/power_cells.cpp

```diff
@@ -0,0 +1,578 @@
+    row->getBBox(bbox);
+
+    row_search.insert({Shape::rectToBox(bbox), row});
+  }
+  auto get_instance_rows = [&row_search] (odb::dbInst* inst) -> std::set<odb::dbRow*> {
+    std::set<odb::dbRow*> rows;
+
+    odb::Rect box;
+    inst->getBBox()->getBox(box);
+
+    for (auto itr = row_search.qbegin(bgi::intersects(Shape::rectToBox(box)));
+         itr != row_search.qend();
+         itr++) {
+      auto* row = itr->second;
+      odb::Rect row_box;
+      row->getBBox(row_box);
+
+      const auto overlap = row_box.intersect(box);
+      if (overlap.minDXDY() != 0) {
+        rows.insert(row);
+      }
+    }
+
+    return rows;
+  };
+
+  for (auto* row : rows) {
+    const int site_width = row->getSite()->getWidth();
+    cell_->populateAlwaysOnPinPositions(site_width);
+    if (row->getOrient() == odb::dbOrientType::R0) {
+      continue;
+    }
+    const std::string inst_prefix = inst_prefix_ + row->getName() + "_";
+    int idx = 0;
+
+    debugPrint(grid_->getLogger(), utl::PDN, "PowerSwitch", 2, "Adding power switches in row: {}", row->getName());
+
+    odb::Rect bbox;
+    row->getBBox(bbox);
+    std::vector<odb::Rect> straps;
+    for (auto itr = targets.qbegin(bgi::intersects(Shape::rectToBox(bbox)));
+         itr != targets.qend();
+         itr++) {
+      const auto& shape = itr->second;
+      straps.push_back(shape->getRect());
+    }
+
+    std::sort(straps.begin(), straps.end(), [](const odb::Rect& lhs, const odb::Rect& rhs) {
+      return lhs.xMin() < rhs.xMin();
+    });
+
+    for (const auto& strap : straps) {
+      const std::string new_name = inst_prefix + std::to_string(idx);
+      auto* inst = odb::dbInst::create(grid_->getBlock(),
+                                       cell_->getMaster(),
+                                       new_name.c_str(),
+                                       true);
+      if (inst == nullptr) {
+        inst = grid_->getBlock()->findInst(new_name.c_str());
+        if (inst->getMaster() != cell_->getMaster()) {
+          grid_->getLogger()->error(utl::PDN,
+                                    221,
+                                    "Instance {} should be {}, but is {}.",
+                                    new_name,
+                                    cell_->getMaster()->getName(),
+                                    inst->getMaster()->getName());
+        }
+      }
+
+      debugPrint(grid_->getLogger(), utl::PDN, "PowerSwitch", 3, "Adding switch {}", new_name);
+
+      const auto locations = computeLocations(strap, site_width);
+      inst->setLocation(*locations.begin(), bbox.yMin());
+      inst->setPlacementStatus(odb::dbPlacementStatus::FIRM);
+
+      const auto inst_rows = get_instance_rows(inst);
+      if (inst_rows.size() < 2) {
+        // inst is not in multiple rows, so remove
+        odb::dbInst::destroy(inst);
+        debugPrint(grid_->getLogger(), utl::PDN, "PowerSwitch", 3, "Removing switch {} since it is not inside two twos.", new_name);
```

Comment:
Does this need a warning?  It seems like no switch is being created where one is needed.